core/src/main/scala/kafka/idempotence/ProducerIdMapping.scala

```diff
@@ -0,0 +1,333 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  * http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.idempotence
+
+import java.io._
+import java.nio.ByteBuffer
+import java.nio.file.Files
+
+import kafka.common.{KafkaException, TopicAndPartition}
+import kafka.log.LogConfig
+import kafka.utils.{Logging, nonthreadsafe}
+import org.apache.kafka.common.errors.{InvalidSequenceNumberException, ProducerFencedException}
+import org.apache.kafka.common.protocol.types._
+import org.apache.kafka.common.utils.Utils
+
+import scala.collection.mutable
+
+private[idempotence] case class PidEntry(seq: Int, epoch: Short, offset: Long)
+private[idempotence] class CorruptSnapshotException(msg: String) extends KafkaException(msg)
+
+object ProducerIdMapping {
```

Comment:
I renamed it because `IdMapping` seemed too generic and cleaned up enough of the internals for git to effectively see it as a rewrite. The basic logic should be in tact, though the previous test cases were broken. It was reviewed before, but very abstractly since it wasn't actually hooked into anything.